main.c

```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+//
+// This boot method:
+//
+// 1. Extracts the ELF binary.
+// 2. Prepares Multiboot information (to pass the optional command line).
+// 3. Jumps to the extracted ELF's entry point.
+void elf_boot(struct boot_params params)
+{
+    die_on(
+        params.kernel_addr == 0,
+        "kernel start address not found!\n"
+        "The VMM does not offer an ELF via fw-cfg. Is the relevant config option missing?\n");
+
+    // Minimal (32bit) ELF loading.
+    struct elf_header *elf = (struct elf_header *)params.kernel_addr;
+    printf("Loading ELF from address: %p\n", elf);
+
+    bool is_elf_executable_or_unknown =
+        elf->type == ELF_TYPE_EXECUTABLE || elf->type == ELF_TYPE_UNKNOWN;
+    die_on(!is_elf_executable_or_unknown, "Unknown ELF type 0x%x!\n", elf->type);
+    die_on(elf->class == 2, "64bit ELF not yet supported!\n");
+    die_on(
+        elf->phentsize != 32,
+        "32bit ELF file should have program header entries with size 32, but found size %x!\n",
+        elf->phentsize);
+
+    // The ELF program header table describes how to prepare the memory layout.
+    printf("ELF program header table offset is 0x%x\n", elf->phoff);
+
+    // The ELF file is loaded to memory already. We need that start addressto copy its contents
+    // (such as code and data) to the correct location.
+    uint8_t *elf_addr_in_memory = (uint8_t *)elf;
+
+    // Each entry in the program header table describes a memory region.
+    // We prepare the contents of each region, but ignore the access rights.
+    struct elf_program_header *current_program_header =
+        (struct elf_program_header *)(elf_addr_in_memory + elf->phoff);
+
+    for (uint16_t i = 0; i < elf->phnum; i++) {
+
+        printf(
+            "Preparing ELF program header (type 0x%x, offset 0x%x, vaddr %p, paddr %p, filesize 0x%x, memsize 0x%x)\n",
+            current_program_header->type, current_program_header->offset,
+            current_program_header->vaddr, current_program_header->paddr,
+            current_program_header->filesize, current_program_header->memsize);
+
+        uint8_t *current_elf_segment_in_memory =
+            elf_addr_in_memory + current_program_header->offset;
+
+        // Pray that nothing is where the ELF contents want to be.
+        // XXX: Is there a way to check it?
+        uint8_t *dest_addr = (uint8_t *)current_program_header->vaddr;
+
+        // Copy contents of the region from the ELF file.
+        memcpy(dest_addr, current_elf_segment_in_memory, current_program_header->filesize);
+
+        // Fill remainder of the region with zeroes (used for BSS segment).
+        memset(dest_addr + current_program_header->filesize, 0,
+               current_program_header->memsize - current_program_header->filesize);
+
+        current_program_header++;
+    }
+
+    // Prepare Multiboot information.
+    struct mb_boot_information_required *mbinfo = malloc(sizeof(*mbinfo));
+    mbinfo->flags = 0;
```

Comment:
`memset` with 0 is a better option here to be sure everything is initialized and no random memory leaks into the structure.

Reply:
Fixed.